docs: Use String in Rc::into_raw examples

vorner opened the pull request:
It is unclear if accessing an integer after `drop_in_place` has been
called on it is undefined behaviour or not, as demonstrated by the
discussion in
https://github.com/rust-lang/rust/pull/60766#pullrequestreview-243414222.

Avoid these uncertainties by using String which frees memory in its
`drop_in_place` to make sure this is undefined behaviour. The message in
the docs should be to watch out and not access the data after that, not
discussing when one maybe could get away with it O:-).

rust-highfive commented:
r? @JoshTriplett

(rust_highfive has picked a reviewer for you, use r? to override)

Centril commented:
r? @RalfJung

RalfJung commented:
Despite what the title says, this is just for the `Weak::*_raw` functions, right? What about the similar functions on `Rc` itself?

vorner commented:
Ups, yes, the title is wrong.

I don't think the original examples had this problem of claiming something is UB and *maybe* not being UB in the first place, so I've changed only the examples I've added in the linked merge request. Do you think it makes sense to go through the other examples too and unify it (for no other reason than just the consistency)?

RalfJung commented:
> Do you think it makes sense to go through the other examples too and unify it (for no other reason than just the consistency)?

Yes I think that would make sense. But if you don't want to do that that's fine too; this PR is okay as-is (if you fix the title).

vorner commented:
I'll have a look and go through the other examples, that's fine (it'll probably take few days before I get around to it, though).

vorner commented:
OK, fixup with more example updates (I'll squash after full review, separating incremental changes for now).

I've done it only with the *_raw methods, not on the `fn new` and similar ‒ the .to_owned() and such feels a bit distracting in the examples. Is that OK?

RalfJung commented:
> I've done it only with the *_raw methods, not on the fn new and similar ‒ the .to_owned() and such feels a bit distracting in the examples. Is that OK?

Yes, that sounds great.

I thought somewhere in these changed examples we'd also get the "but not anymore this, accessing the memory now would be UB", but I guess that can only happen with `Weak`?

vorner commented:
> I thought somewhere in these changed examples we'd also get the "but not anymore this, accessing the memory now would be UB", but I guess that can only happen with Weak?

There's this one example that has a dangling pointer at the end (because the Arc was already freed), but that is dangling with an int inside it too. This specific UB can happen only with Weak, yes.

Should I rebase? I've looked at the failing CI, but my feeling is this is problem in the build itself?

RalfJung commented:
Ignore Azure CI, that is just being tested currently.

RalfJung commented:
> This specific UB can happen only with Weak, yes.

Fair. All right then, let's land this.

@bors r+

bors commented:
:pushpin: Commit e5af0e51ab1e436fda8f8a45115114a46deebf0d has been approved by `RalfJung`

<!-- @bors r=RalfJung e5af0e51ab1e436fda8f8a45115114a46deebf0d -->
<!-- homu: {"type":"Approved","sha":"e5af0e51ab1e436fda8f8a45115114a46deebf0d","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit e5af0e51ab1e436fda8f8a45115114a46deebf0d with merge ab0c4b9cd9b2f59b033abfc13082bd76cc849a5e...
<!-- homu: {"type":"BuildStarted","head_sha":"e5af0e51ab1e436fda8f8a45115114a46deebf0d","merge_sha":"ab0c4b9cd9b2f59b033abfc13082bd76cc849a5e"} -->

vorner commented:
Um, wait? I was talking about rebasing because of the fixup commit that needs to be squashed before the merge (I've pushed the rebase now). Or does bors handle that automatically? It would be stupid if the fixup commit ended up in master :-(.

RalfJung commented:
Sorry, I didn't notice the fixup commit.

@bors retry

pietroalbini commented:
@bors r-

RalfJung commented:
@pietroalbini why that?

pietroalbini commented:
Based on @vorner's comment this wasn't supposed to land until the fixup was removed, but `@bors retry` doesn't do that.

...I now see the commit was already pushed auto-unapproving it...

RalfJung commented:
Oh that's what happened. I did the retry to cancel the build that was already ongoing. I forgot about auto-unapproval. (Is that state change visible anywhere from within the issue?)

@bors r+

bors commented:
:pushpin: Commit 79e58399920d6a0a7c4050bfb26c64fbc10562ec has been approved by `RalfJung`

<!-- @bors r=RalfJung 79e58399920d6a0a7c4050bfb26c64fbc10562ec -->
<!-- homu: {"type":"Approved","sha":"79e58399920d6a0a7c4050bfb26c64fbc10562ec","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit 79e58399920d6a0a7c4050bfb26c64fbc10562ec with merge fc550d4295a654f9e7c621d957d81fbf1426c70b...
<!-- homu: {"type":"BuildStarted","head_sha":"79e58399920d6a0a7c4050bfb26c64fbc10562ec","merge_sha":"fc550d4295a654f9e7c621d957d81fbf1426c70b"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/115580653), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/25283168)
Approved by: RalfJung
Pushing fc550d4295a654f9e7c621d957d81fbf1426c70b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang/rust/builds/25283168","checks-travis":"https://travis-ci.com/rust-lang/rust/builds/115580653"},"merge_sha":"fc550d4295a654f9e7c621d957d81fbf1426c70b"} -->
